pilot/pkg/model/push_context.go

```diff
@@ -310,15 +312,50 @@ func (ps *PushContext) UpdateMetrics() {
 }
 
 // Services returns the list of services that are visible to a Proxy in a given config namespace
-func (ps *PushContext) Services(proxy *Proxy) []*Service {
+func (ps *PushContext) Services(p *Proxy) []*Service {
 	// TODO: use network scopes here, and return services in the proxy namespace if
 	// configured to use current_namespace as the scope and proxy is not nil
-	//if proxy == nil {
-	//	return ps.allServices
-	//}
+	if p == nil {
+		return ps.allServices
+	}
+	p.mutex.RLock()
+	defer p.mutex.RUnlock()
+	if p.serviceDependencies != nil {
+		return p.serviceDependencies
+	}
 	return ps.allServices
 }
 
+// OnConnect is called when a node connects. Will update per-node data.
+func (ps *PushContext) OnConnect(p *Proxy) {
+	// For now Router (Gateway) is not using the isolation - the Gateway already has explicit
+	// bindings.
+	if pilot.NetworkScopes != "" && p.Type == Sidecar {
+		// Add global namespaces. This may be loaded from mesh config ( after the API is stable and
+		// reviewed ), or from an env variable.
+		adminNs := strings.Split(pilot.NetworkScopes, ",")
+		globalDeps := map[string]bool{}
+		for _, ns := range adminNs {
+			globalDeps[ns] = true
+		}
+
+		p.mutex.RLock()
+		defer p.mutex.RUnlock()
+		res := []*Service{}
+		for _, s := range ps.allServices {
+			podNamespace := s.Attributes.Namespace
+			if podNamespace == "" {
+				res = append(res, s)
+			} else if globalDeps[podNamespace] || podNamespace == p.ConfigNamespace {
+				res = append(res, s)
```

Comment:
This logic is incorrect. It does not take into account changes to the list of services. Secondly, given that we don't have some onDemand logic, at this moment, having the logic here only adds overhead instead of optimizing anything. I don't see how this is optimized (today) than simply having a single shared map in the PushContext that is of the form
 `Namespace->listOfServicesThatThisNamespaceDependsUpon`
This list can be updated anytime services are added/removed. You compute this list ONLY ONCE when initializing the push context. Not for every proxy that connects. You don't have to track 1000s of proxy objects and trigger recomputation of the dependent services, and other configuration artifacts when services are added/removed from the service registry or configuration objects are added/removed/modified. Its fine to have logic that will be used in future, but not when this logic has poorer performance than what is there today.